sdk/helper/useragent/useragent.go

```diff
@@ -26,22 +27,39 @@ var (
 // String returns the consistent user-agent string for Vault.
 //
 // e.g. Vault/0.10.4 (+https://www.vaultproject.io/; go1.10.1)
-func String() string {
-	return fmt.Sprintf("Vault/%s (+%s; %s)",
-		versionFunc(), projectURL, rt)
+//
+// Given comments will be appended to the semicolon-delimited comment section.
+//
+// e.g. Vault/0.10.4 (+https://www.vaultproject.io/; go1.10.1; comment-0; comment-1)
+func String(comments ...string) string {
+	c := strings.Join(comments, "; ")
+	if len(c) > 0 {
+		c = "; " + c
+	}
+
+	return fmt.Sprintf("Vault/%s (+%s; %s%s)",
+		versionFunc(), projectURL, rt, c)
 }
```

Comment:
Total nit here but perhaps we could use `strings.Join` to construct the entire string here, something like:
```suggestion
	comment = append([]string{projectURL, rt}, comment...)
	c := strings.Join(comment, "; ")

	return fmt.Sprintf("Vault/%s (+%s)", versionFunc(), c)
}
```

Reply:
Forgot to mention, but similar for `PluginString` as well.

Reply:
Ya, I wrote up something using your suggested approach as an alternative. I'd be happy to change things if you think that's better.

Here is what I wrote for `String()`, which is similar to your suggestion:
```
func String(comments ...string) string {
	c := append([]string{"+" + projectURL, rt}, comments...)
	return fmt.Sprintf("Vault/%s (%s)", versionFunc(), strings.Join(c, "; "))
}
```

For `PluginString()`, the `pluginName` could be empty, so it doesn't end up being as clean. I wonder if it's important to preserve the order of the comments in case something depends on it. For example, a server could be splitting/indexing on the semicolons. The placement of `c = append(c, rt)` below preserves the original order where `pluginName` comes first in the comments. 
```
func PluginString(env *logical.PluginEnvironment, pluginName string, comments ...string) string {
	c := []string{"+" + projectURL}
	if pluginName != "" {
		c = append(c, pluginName)
	}
	c = append(c, rt)
	c = append(c, comments...)
	return fmt.Sprintf("Vault/%s (%s)", env.VaultVersion, strings.Join(c, "; "))
}
```

Let me know your thoughts!

Reply:
I slightly prefer the versions in your comments. It's slightly cleaner to read and it's nice that the comments are all logically grouped together in a way that aligns very closely with the [RFC spec](https://httpwg.org/specs/rfc7231.html#header.user-agent).

Reply:
+1 - I like that all of the comments are grouped together. I've changed this in e8df12f.